.circleci/config.yml

```diff
@@ -22,21 +31,198 @@ anchors:
+    steps:
+      - run:
+          command: |
+            if [ ${CIRCLE_BRANCH} != "" ] &&
+              [ ${CIRCLE_BRANCH} != "master" ] &&
+              [ $(git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep << parameters.dir_name >>/ | wc -l | xargs) == 0 ]; then
+                echo "No relevant changes in << parameters.dir_name >> directory in non-master branch. Skipping."
+                circleci step halt
+            fi
+          name: Ensure branch has changes to API or UI code
+
+  ui_dependencies:
+    description: "workbench/ui: yarn install, save and restore cache"
+    parameters:
+      version:
+        type: string
+        default: << pipeline.parameters.cache_key_version >>
+    steps:
+      # caching: https://circleci.com/docs/2.0/caching/
+      - restore_cache:
+          keys:
+            # restore cache strategy:
+            # Check for the latest cache from your branch.
+            # If that fails, then check for the latest from master branch.
+            # If that fails, then check for the latest from any branch
+            - << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}-{{ .Branch }}
+            - << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}-master
+            - << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}
+      - run:
+          name: "workbench/ui: yarn install dependencies"
+          working_directory: ~/workbench/ui
+          command: yarn install --verbose --frozen-lockfile --non-interactive
+      - save_cache:
+          paths:
+            - ~/.cache/yarn
+            - ui/node_modules
+          key: << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}-{{ .Branch }}
+
+  e2e_dependencies:
+    description: "workbench/e2e: yarn install, save and restore cache"
+    parameters:
+      version:
+        type: string
+        default: << pipeline.parameters.cache_key_version >>
+    steps:
       - restore_cache:
           keys:
-          - api-cache-{{ checksum "~/workbench/api/build.gradle" }}
-          - api-cache-
+            - << parameters.version >>-yarn-e2e-cache-{{ checksum "~/workbench/e2e/yarn.lock" }}-{{ .Branch }}
+            - << parameters.version >>-yarn-e2e-cache-{{ checksum "~/workbench/e2e/yarn.lock" }}-master
+            - << parameters.version >>-yarn-e2e-cache-{{ checksum "~/workbench/e2e/yarn.lock" }}
+      - run:
+          name: "workbench/e2e: yarn install dependencies"
+          working_directory: ~/workbench/e2e
+          command: yarn install --frozen-lockfile --non-interactive
+      - save_cache:
+          paths:
+            - ~/.cache/yarn
+            - e2e/node_modules
+          key: << parameters.version >>-yarn-e2e-cache-{{ checksum "~/workbench/e2e/yarn.lock" }}-{{ .Branch }}
+
+  api_cache:
+    description: "workbench/api: save and restore gradle cache"
+    parameters:
```

Comment:
So if they're both false, then nothing happens. It seems like one is required for meaningful operation. An accidental no-op command invocation could sit unnoticed for some time.

You could do a `cache_mode` parameter with an enum like `[SAVE, RESTORE, SAVE_AND_RESTORE]` and make that required? That would solve the potential no-op problem (though if you add a third command operation, it would get complicated).

Finally, you could break this down into very simple commands like `save_api_cache` and `restore_api_cache`, or even a `save_cache` with the name of the cache as a parameter.